### PR TITLE
AirfRANS: T_max=10 phase transition deep dive with LR decay

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS shows a dramatic phase transition at epoch ~40 with T_max=10 — val surface_mse drops sharply at the cosine trough (PR #2556, emma, W&B: 7qre8z5x). This is the most striking result in the AirfRANS experiments so far.

The hypothesis: after the phase transition at epoch 40, the model enters a higher-quality basin. If we then apply a decaying LR (multiply the cosine peak by 0.7 each cycle), subsequent cycles stay in this basin and refine it more precisely rather than escaping via high-LR re-exploration.

Concretely: use `--cosine_t_max 10` as baseline but add `--cosine_decay 0.7` so each cosine peak is 70% of the previous one. This keeps the beneficial phase-transition structure but progressively tightens the search radius as the model converges — similar to simulated annealing.

If `--cosine_decay` is not yet a supported flag, implement it by modifying the LR scheduler: after each T_max cycle completes, multiply `eta_max` by 0.7 before starting the next cycle.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 10 \
  --cosine_decay 0.7 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --epochs 999 \
  --wandb_group airfrans-tmax10-phase-transition
```

Key change vs baseline: `--cosine_decay 0.7` — reduces LR peak by 30% each cosine cycle, creating a decaying envelope over the oscillating schedule.

If `--cosine_decay` is not a supported argument, use `CosineAnnealingWarmRestarts` with `T_mult=1` and manually reduce `eta_max` by 0.7 after each restart in the training loop.

All other flags match the current AirfRANS best configuration (PR #2556).

## Baseline

- **Dataset:** AirfRANS
- **val surface_mse (primary metric):** 0.0248
- **Baseline W&B run:** 7qre8z5x
- **Baseline PR:** #2556 (emma)
- **Reproduce command:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 10 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing — LR decay without grad-clip is obsolete. Baseline now 0.01419 via grad-clip=1.0. armin, please pick up your next assignment.

**Comment by tcapelle** [comment]
STUDENT armin:

## Results

**Outcome: NEGATIVE** — cosine_decay=0.7 killed the learning rate too quickly, preventing the phase transition that makes T_max=10 effective.

### Key Metrics

| Metric | Baseline (PR #2556) | This experiment | Delta |
|--------|---------------------|-----------------|-------|
| val_primary/surface_mse | 0.0248 | 0.12362 | +398% (5x worse) |
| full_val/surface_mse_p | — | 0.4934 | — |
| full_val/volume_mse | — | 0.2504 | — |

- **Best val:** 0.12362 at epoch 46 (flatlined from epoch 20 onward)
- **W&B run:** `1rjc6g61`
- **Peak GPU memory:** ~23.9 GB (GPU 1)
- **Epochs:** 149+ (still running to 180-min timeout, but result finalized at epoch ~20)

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 10 \
  --cosine-decay 0.7 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --epochs 999 \
  --agent armin \
  --wandb-name "armin/airfrans-tmax10-cosine-decay07" \
  --wandb-group airfrans-tmax10-phase-transition
```

### What happened

The per-epoch decay of 0.7 compounds catastrophically with the per-batch cosine schedule. The LR trajectory:

| Epoch | LR | val_primary/surface_mse |
|-------|------|-------------------------|
| 1 | 3.5e-4 | 0.392 |
| 5 | 8.4e-5 | 0.237 |
| 10 | 1.4e-5 | 0.139 |
| 20 | 4.0e-7 | 0.124 |
| 40 | 3.2e-10 | 0.124 |
| 46 (best) | 3.8e-11 | 0.124 |
| 149 | 8.5e-27 | 0.124 |

The decay factor 0.7^epoch means LR drops to 2.8% of peak by epoch 10 and is essentially zero by epoch 20. The baseline's phase transition happens at epoch ~40 (where T_max=10's cosine trough creates a deep basin), but by that point the decayed LR is 10^-10 — the model can't move at all.

The core problem: **the cosine decay destroys the very mechanism that makes T_max=10 work.** The phase transition requires the model to traverse sharp LR peaks (full lr at cosine peaks) to escape local minima and then settle into deep basins at troughs. With decay=0.7/epoch, by epoch 10 the "peaks" are already negligible, and the model gets permanently stuck in whatever early-convergence plateau it found (val=0.124).

### Suggested follow-ups

1. **Much slower decay rates** (0.95 or 0.99/epoch) — the original hypothesis of refinement after phase transition could work, but only if the decay is slow enough to preserve the first 40+ epochs of full-amplitude cycling.
2. **Delayed decay onset** — e.g. no decay for first 50 epochs, then 0.9/epoch. This would let the phase transition happen naturally, then progressively tighten.
3. **Step-decay at specific milestones** — e.g. cut LR by 0.5× at epoch 100, 200. More targeted than continuous decay.
4. The fact that lr=2e-4 + T_max=50 (PR #2580) reached 0.00300 over 240 epochs without any decay suggests that aggressive LR scheduling may not be needed — longer training with consistent amplitude may be the better strategy for AirfRANS.
